src/prognostic_equations/implicit/autodiff_utils.jl

```diff
@@ -0,0 +1,80 @@
+end
+
+# An iterator over the FieldNames of all scalar fields in a FieldVector.
+scalar_field_names(field_vector) =
+    MatrixFields.filtered_names(field_vector) do x
+        x isa Fields.Field && eltype(x) == eltype(field_vector)
+    end
+
+# An iterator with pairs of the form (scalar_level_index, (name_index, level)),
+# where name_index is an index into scalar_field_names(field_vector), level is
+# a vertical index into the scalar field with this name, and scalar_level_index
+# is a unique linear index for each combination of scalar field name and level.
+function scalar_level_pair_iterator(field_vector)
+    scalar_names = scalar_field_names(field_vector)
+    name_index_and_level_pairs =
+        unrolled_map(enumerate(scalar_names)) do (name_index, name)
+            field = MatrixFields.get_field(field_vector, name)
+            is_one_level = field isa Fields.SpectralElementField
+            Iterators.map(
+                level -> (name_index, level),
+                Base.OneTo(is_one_level ? 1 : Spaces.nlevels(axes(field))),
+            )
+        end
+    return enumerate(Iterators.flatten(name_index_and_level_pairs))
+end
+
+# A view of the data for one point in a field, which can be used like a Ref.
+Base.@propagate_inbounds function point(field, level, column_index...)
+    column_field = Fields.column(field, column_index...)
+    column_field isa Fields.PointField && return column_field
+    first_level_index = Operators.left_idx(axes(column_field))
+    return Fields.level(column_field, first_level_index + level - 1)
+end
+
+# TODO: This needs to be moved into ClimaCore to avoid breaking precompilation.
+# function ClimaComms.context(x::Fields.FieldVector)
+#     values = Fields._values(x)
+#     isempty(values) && error("Empty FieldVector has no device or context")
+#     index = unrolled_findfirst(Base.Fix2(!isa, Fields.PointField), values)
+#     return ClimaComms.context(values[isnothing(index) ? 1 : index])
+# end
+
+##################################################
+## Nonstandard rules for automatic differentiation
+##################################################
+
+# Every nonstandard rule should be defined using a method that specializes on
+# the tag `Jacobian`. Not specializing on any tag overwrites the generic method
+# for `Dual` in `ForwardDiff` and breaks precompilation, while specializing on
+# the default tag `Nothing` causes the type piracy test to fail.
+
+# Set the derivative of `sqrt(x)` to `iszero(x) ? zero(x) : inv(2 * sqrt(x))` in
+# order to properly handle derivatives of `x * sqrt(x)`. Without this change,
+# the derivative of `x * sqrt(x)` is `NaN` when `x` is zero.
+@inline function Base.sqrt(d::ForwardDiff.Dual{Jacobian})
+    tag = Val(Jacobian)
+    x = ForwardDiff.value(d)
+    partials = ForwardDiff.partials(d)
+    val = sqrt(x)
+    deriv = iszero(x) ? zero(x) : inv(2 * val)
+    return ForwardDiff.dual_definition_retval(tag, val, deriv, partials)
+end
```

Comment:
I understand that you need to avoid NaN for numerical purposes, but how do we know that replacing them with 0 does not change Newton's iteration solution?